### ISSUE TITLE
Cache key clear process can consume too many CPU

### ISSUE THREAD

**Issue by turgayozgur** [nopSolutions/nopCommerce#3475]
https://github.com/nopSolutions/nopCommerce/blob/b16776c35972faf7b72d47bd11197264030874c4/src/Libraries/Nop.Core/Caching/MemoryCacheManager.cs#L127

Trying to clean cache keys for every eviction can consume too many cpu resource if you use nopcommerce for an app that under high request and too many cache key depends on own implementation.

**Comment by antalistvan** [comment]
Same issue here... After a "Clear Cache" or after any Category modification on the Admin pages, the site can stuck in an unresponsive state for minutes, using an F8 size VM in Azure. Something need to do with it.

**Comment by soci1248** [comment]
Can someone explain what is the rationale behind the MemoryCacheManager?
1. What is the exact responsibilities it wants to achieve?
2. Particularly, what is the purpose of the _allKeys dictionary? Why it is needed, why the underlying cache cannot supply this information?
I would be happy to optimize it and contribute it back, but first I have to understand exactly why did you introduce the _allKeys dictionary? For larger sites it kills all the cpus, this is a serious performance bug.

**Comment by turgayozgur** [comment]
@soci1248 
1. To manage the memory caching by single point and be able to make changes at root level if required. Also it manage the clearing cache keys by pattern and abstract the usage to simplify that.
2. _allKeys dic is responsible for the deleting cache object by regex pattern somthing like starts with given key.

**Comment by soci1248** [comment]
Based on a preliminary analysis I see the following problems. 
When there a many items in the cache RegisterPostEvictionCallback calls back the PostEviction method very frequently. Consider 100000 items, all expiring in 1 hour, and they are evenly distributed with an hour. Then every 3600 sec / 100000=0.036 second you got a callback in this method. This is 27 call per second. The real situation is much worse, because the cache additions are not evenly distributed at all in time.
What does PostEviction do? It calls ClearKeys, which iterates thru _allKeys using a linear search (Where). For 100000 items this needs considerable time and CPU. This is the reason a production server is experiencing slowdown during a day. I used procdump to create a memory dump when the CPU for the worker process is more than 90% for 30 seconds. What did I see in the dump? 14 of the 30 threads were executing ClearKeys! Because it is a tight CPU loop, it sucks all CPUs to the ground.

Why do you call ClearKeys for when an item is removed from the cache? Why bother checking other items during the callback of a particular item? Why not call ClearKeys on a timer thread for every 10 seconds or so?

The next one, this actually also causes serious CPU burning is the IsCrawler method. It creates 5000 regex object for each call, and evaluates all of them if the user agent string is not a search engine. If you simply create and store the Regex object instead of string during the initialization the performance doubles at least. If you concat the pieces by (regex1)|(regex2), etc. manner to create one, big regex, the performance becomes even better.

Why I mention this method too? After a day of profiling I can claim that 75% of the CPU is eaten by ClearKeys and IsCrawler!!!

One other problem which does not bother me at this moment because I don't see it in the profiler and the memory dumps, but can cause pain in the back later: RemoveByPattern. 
This is problematic, the code creates a _compiled_ regex for each call! Compiled regex without caching the regex objects? 
https://stackoverflow.com/questions/513412/how-does-regexoptions-compiled-work

**Comment by turgayozgur** [comment]
> Based on a preliminary analysis I see the following problems.
> When there are many items in the cache RegisterPostEvictionCallback calls back the PostEviction method very frequently. Consider 100000 items, all expiring in 1 hour, and they creation are evenly distributed within an hour. Then every 3600 sec / 100000=0.036 second you got a callback in this method. This is 27 call per second. The real situation is much worse, because the cache additions are not evenly distributed at all in time.
> What does PostEviction do? It calls ClearKeys, which iterates thru _allKeys using a linear search (Where). For 100000 items this needs considerable time and CPU. This is the reason a production server is experiencing slowdown during a day. I used procdump to create a memory dump when the CPU for the worker process is more than 90% for 30 seconds. What did I see in the dump? 14 of the 30 threads were executing ClearKeys! Because it is a tight CPU loop, it sucks all CPUs to the ground.
> 
> Why do you call ClearKeys for when an item is removed from the cache? Why bother checking other items during the callback of a particular item? Why not call ClearKeys on a timer thread for every 10 seconds or so?
> 
> The next one, this actually also causes serious CPU burning is the IsCrawler method. It creates 5000 regex object for each call, and evaluates all of them if the user agent string is not a search engine. If you simply create and store the Regex object instead of string during the initialization the performance doubles at least. If you concat the pieces by (regex1)|(regex2), etc. manner to create one, big regex, the performance becomes even better.
> 
> Why I mention this method too? After a day of profiling I can claim that 75% of the CPU is eaten by ClearKeys and IsCrawler!!!
> 
> One other problem which does not bother me at this moment because I don't see it in the profiler and the memory dumps, but can cause pain in the back later: RemoveByPattern.
> This is problematic, the code creates a _compiled_ regex for each call! Compiled regex without caching the regex objects?
> https://stackoverflow.com/questions/513412/how-does-regexoptions-compiled-work

That is what I exactly say.:) But no one from the nop team, discuss the issue, until someone facing to this.

**Comment by fpdavis** [comment]
> The next one, this actually also causes serious CPU burning is the IsCrawler method. It creates 5000 regex object for each call, and evaluates all of them if the user agent string is not a search engine. If you simply create and store the Regex object instead of string during the initialization the performance doubles at least. If you concat the pieces by (regex1)|(regex2), etc. manner to create one, big regex, the performance becomes even better.

I modified the BrowscapXmlHelper.cs as suggested above and placed it in a test harness. I can confirm that this will provide a boost in performance to this method. On average the isCrawler() method ran in about 1/3rd the time. I also tried using the RegexOptions.Compiled but I verified that there is a bug in newer versions of .Net which actually causes a significant **decrease** in performance when using this option.

My test harness with updated BrowscapXmlHelper code is available at https://github.com/fpdavis/BrowsCapTest

**Comment by soci1248** [comment]
Good progress, thanks. One more very easy but dramatically efficient optimization could exploit the fact that there is a strong serial correlation between the bot requests. I.e. google starts to crawl a site it will do it for the next several minutes. When the BrowscapXmlHelper identified a bot using the slow regex approach it could cache the exact user agent string which come from the bot. So, for the subsequent request it could answer from this cache within nanoseconds.
A HashSet + lock might be appropriate for this task, or you could use ConcurrentDictionary with null Value.
If you afraid filling the memory with cached user agent strings you can use a sliding time expiration cache with 1-2 hours of duration.

**Comment by soci1248** [comment]
What about the ClearKeys()? As I wrote it causes 3 times more severe performance problems than the regex thing. What I don't understand from the design, why do you store the keys redundantly when the underlying cache already stores them?
It seems maintaining this redundancy is the reason of the big perf hit.

Actually, I don't understand the intent of the developer.
                //try to remove key from dictionary
                if (!_allKeys.TryRemove(key, out _))
                    //if not possible to remove key from dictionary, then try to mark key as not existing in cache
                {
                    _allKeys.TryUpdate(key, false, true);
                }

When TryRemove returns false? When the item is already removed by a concurrent thread. Then why do you want to update it to false when it is already deleted?
It is possible I misunderstood the code, but I think some serious thinking is needed for this implementation. 
After you make it efficient, you can advertise the product become 4 times faster. :)

**Comment by antalistvan** [comment]
Currently, we are trying to workaround the ClearKeys issue.

We made the following changes in the MemoryCacheManager.cs:

Added a new class property, contains only those keys are marked to delete:
`        protected static readonly ConcurrentDictionary<string, bool> _keysToRemove;`

In the CTOR we initialized that
`           _keysToRemove = new ConcurrentDictionary<string, bool>();`

Changed the TryRemoveKey's inner content to this:

```
            _keysToRemove.TryRemove(key, out _);
            if (!_allKeys.TryRemove(key, out _))
            {
                if (_allKeys.TryUpdate(key, false, true))
                {
                    _keysToRemove.TryAdd(key, true);
                }
            }
```

And finally replace the ClearKeys to this:
```
            foreach (var key in _keysToRemove.Keys.ToList())
            {
                TryRemoveKey(key);
            }
```

We are testing this modification in our production environment, and after an hour, we experienced that the overall site performance is much better than anytime before.

It's not 20% or 50%. It's about 2-3x faster than before.

For example a simple ajax call went down from about 500ms to 100-130ms.

**Comment by turgayozgur** [comment]
@antalistvan The better way is just calling the ClearKeys method by given time period. For example, use a timer that ticks every 10 seconds.

Something like that:
`var timer = new Timer(state => ClearKeys(), context, WAIT_TIME_TO_START, period);`

**Comment by soci1248** [comment]
Burning the CPU even every 10 seconds is a waste of time yet. This issue should be handled by correct data structures, not by slowing down the CPU sucking cycle. _keysToRemove idea is much better i believe.

**Comment by turgayozgur** [comment]
It will not burn your cpu if you have lower than ~1 million cache key. The issue not about calling the ClearKeys. It is about calling this method too many times for a second. Actually, ClearKeys method not must. Ignoring this method is another solution I think.

**Comment by soci1248** [comment]
Sure, less call, less cpu.
Does someone who wrote the code can reason about my TryRemove opinion? And why we need a redudant key collection when the underlying cache already stores them?

**Comment by antalistvan** [comment]
Just another proof.. From the picture, is clearly seeable the modification make the site performs better.
![site avaibility](https://user-images.githubusercontent.com/22097078/53796668-e7201700-3f34-11e9-9869-a79077ee391d.png)

**Comment by fpdavis** [comment]
So a quick update on this. We were having server issues every day during peak load where multiple servers would slow down to a crawl until we restarted it. After implementing the isCrawler Regex change we have not had any server issues. We are also looking at the ClearKeys but we have a few unanswered questions around that before we figure out the best way to address it (as others have been pointing out). We will post more on this later.

As far as caching the exact user agent string from the request and checking the cache first before running the string through the Regex... I was a little confused why it would ever need to run the customer back through the isCrawler method once the "customer" was identified as a bot. After further examination of the code I see that it does indeed run every non cached, non System account through this check.

My question is why isn't the call to _authenticationService.GetAuthenticatedCustomer() above the IsSearchEngine check in WebWorkContext.cs? This would drastically reduce the number of times the isCrawler method is called.

**Comment by SuperArtie** [comment]
Just to chime in here, I've seen drastic improvement in performance/sitespeed just by implementing the suggested changes to **BrowseXmlHelper** and **IsCrawler()** that @soci1248 outlined.

I took the 4000+ strings used for detecting a site-crawler (each one instantiating a new RegEx), and I concatenated them all to 1 giant, pipe-delimited string, and then instantiated 1 big RegEx from that, which has resulted in just 1 RegEx.IsMatch() for every request's Http User Agent.

I went from having multiple machines in a server-farm having their CPUs pegged (causing the site to crash) to having consistent sitespeed, which ranges between average load times (pretty good, but not great) to somewhat slow (but still responsive) load times. The bottom line though is that having a usable site is a step forward for me and my team.

I've yet to explore the potential gains from refactoring **MemoryCacheManager** and **ClearKeys()* as @antalistvan has outlined, but will report back the findings when I do.

**Comment by soci1248** [comment]
Good news. :)
When you implement all changes we suggested you will trim that 70% overhead entirely (about 3-fold perf gain). Your CPUs become idle, and your customers happy. :)

**Comment by Durkee** [comment]
Hi @skoshelev,
Do you have an idea on when these check ins will occur (i.e. MemoryCacheManager and BrowsCapHelper)?
Thank you for all you do!

**Comment by skoshelev** [comment]
Hi all. Thank you for help us and all of your investigation, I starting to refactoring of MemoryCacheManager class and try to deal with this problem.
As for the reasons for duplicating the list of keys and the whole approach that causes you questions and our performance problems:  when upgrading to version 4.0, we faced the fact that the base IMemoryCache interface does not contain methods or properties that allow access to the list of keys, but we need it for deleting items by mask

**Comment by skoshelev** [comment]
Thanks again to all for your help. In the end, we decided to use EasyCaching.IsoSachingProvider instead of IMemoryCache

Closed #3475

**Comment by skoshelev** [comment]
Slow performance when the cached items count reaches cache size limit in the EasyCaching

depends on https://github.com/dotnetcore/EasyCaching/issues/185

Details in the ticket 15162 in the premium support

**Comment by skoshelev** [comment]
Closed #3475
